#  Classes numériques :

La classe numérique permet de construire les réels et les entiers avec un certain niveau de précision.

- Les entiers (8 classes) : int\* et uint\*
- Les flottants (2 classes) : single (32-bits), double (64-bits)

##  Les Entiers :

Matlab possède quatre classes entières signées et quatre classes entières non signées. Les types signés vous permettent de travailler aussi bien avec des nombres entiers négatifs que positifs, mais ne peuvent pas représenter une gamme de nombres aussi large que les types non signés car un bit est utilisé pour désigner un signe positif ou négatif pour le nombre. Les types non signés vous donnent un plus grand choix de nombres, mais ces nombres ne peuvent être que zéro ou positifs.
Matlab prend en charge le stockage de données entières sur 1, 2, 4 et 8 octets. Vous pouvez économiser de la mémoire et du temps d'exécution pour vos programmes si vous utilisez le plus petit type d'entier qui contient vos données. Par exemple, vous n'avez pas besoin d'un entier de 32 bits pour stocker la valeur 100.
Voici les huit classes entières, la plage de valeurs que vous pouvez stocker avec chaque type, et la fonction de conversion Matlab requise pour créer ce type :

|       Classe        |   Plage de valeurs  | Fonction de conversion
| :-----------------: | :-----------------: | :--------------------:
| Entier 8-bits signé | De $-2^7$ à $2^7-1$ | int8
| Entier 16-bits signé | De $-2^{15}$ à $2^{15}-1$ | int16
| Entier 32-bits signé | De $-2^{31}$ à $2^{31}-1$ | int32
| Entier 64-bits signé | De $-2^{63}$ à $2^{63}-1$ | int64
| Entier 8-bits non signé | De 0 à $2^8-1$ | uint8
| Entier 16-bits non signé | De 0 à $2^{16}-1$ | uint16
| Entier 32-bits non signé | De 0 à $2^{32}-1$ | uint32
| Entier 64-bits non signé | De 0 à $2^{64}-1$ | uint64


<u>Valeurs les plus grandes et les plus petites pour les classes entières : </u>

Pour chaque type de données entier, il y a un plus grand et plus petit nombre que vous pouvez représenter avec ce type. Le tableau indique les valeurs les plus grandes et les plus petites pour chaque type de données entières dans la colonne "Plage de valeurs".
Vous pouvez également obtenir ces valeurs avec les fonctions intmax et intmin : 

In [1]:
intmax('int8')

ans = 127


In [2]:
intmin('int8')

ans = -128


Si vous convertissez un nombre dont la valeur est supérieure à la valeur maximale d’un type de données entier, Matlab lui attribue la valeur maximale de ce type. De même, si vous convertissez un nombre inférieur à la valeur minimale du type de données entier, Matlab lui attribue la valeur minimale. Par exemple :

In [3]:
x = int8(300)

x = 127


In [4]:
x = int8(-300)

x = -128


Aussi, lorsque le résultat d'une opération arithmétique impliquant des entiers dépasse la valeur maximale (ou minimale) du type de données, MATLAB la fixe à la valeur maximale (ou minimale) :

In [5]:
x = int8(100) * 3

x = 127


In [6]:
x = int8(-100) * 3

x = -128


## Les flottants : 

En cours de Complément Mathématique 1, nous avons vu que la famille des réels $\mathbb{R}$ est infinie. Cette infinité ne peut pas être entièrement représentée en informatique, c'est pourquoi nous avons recours à un type intermédiaire : les flottants. Matlab représente donc les nombres réels en format flottant double-précision ou simple-précision. La valeur par défaut est la double précision, mais vous pouvez faire n'importe quel nombre avec une simple fonction de conversion.

### Simple précision : 

Toute valeur stockée en tant que valeur unique nécessite 32 bits, formatés comme indiqué dans le tableau ci-dessous :

|       Bits       |   Usage
| :-----------------: | :-----------------:
| 31 | Bit de signe (0 = + et 1 = -)
| 30 à 23 | Bits d'exposants (binaire jusqu'à 127)
| 22 à 0 | bits de mantisse (1.f)

L’intervalle pour la simple précision est de  -3.40282e+38 à -1.17549e-38 et de 1.17549e-38 à 3.40282e+38 et la fonction de conversion est *single*.


### Double précision : 

Toute valeur stockée en double nécessite 64 bits, formatés comme indiqué dans le tableau ci-dessous :

|       Bits       |   Usage
| :-----------------: | :-----------------:
| 63 | Bit de signe (0 = + et 1 = -)
| 62 à 52 | Bits d'exposants (binaire jusqu'à 1023)
| 51 à 0 | bits de mantisse (1.f)


L’intervalle pour la double précision est de  1.79769e+308 à -2.22507e-308 et de  2.22507e-308 à 1.79769e+308 et la fonction de conversion est *double*.

**<u>NB :</u>** Parce que Matlab stocke les nombres de type simple en utilisant 32 bits, ils nécessitent moins de mémoire que les nombres de type double, qui utilisent 64 bits. Cependant, parce qu'ils sont stockés avec moins de bits, les nombres de type simple sont représentés avec moins de précision que les nombres de type double.

### -Inf et Inf 

Dans la définition des flottants, nous incluons aussi deux éléments : -Inf et Inf soit moins l'infini et l'infini. Ainsi, lorsque l'on dépasse la capacité des flottants, Matlab vous retourne -Inf ou Inf comme vous pouvez le constater ci-dessous :


In [8]:
exempleInfini = 1.79771e+308

exempleInfini =  Inf


Pour plus d'informations sur les flottants, vous pouvez consulter les références **[1-2]** ci-dessous. 

# Bibliographie :

**[1]** Michael Overton - *Numerical Computing with IEEE Floating Point Arithmetic* - http://sistemas.fciencias.unam.mx/~gcontreras/joomla15/tmp/Numerical_Computing_with_IEEE_Floating_Point_Arithmetic.pdf

**[2]** Jean-Michel Muller - *Handbook of Floating-Point Arithmetic* - https://www.springer.com/us/book/9780817647056 (Accessible via ENT --> Supernova)